Firestore/core/test/firebase/firestore/util/executor_std_test.cc

```diff
@@ -0,0 +1,251 @@
+  schedule.Push(3, start_time + chr::milliseconds(1));
+
+  EXPECT_FALSE(schedule.PopIfDue(&out));
+  std::this_thread::sleep_for(chr::milliseconds(5));
+
+  EXPECT_EQ(TryPop(), 3);
+  EXPECT_EQ(TryPop(), 2);
+  EXPECT_EQ(TryPop(), 1);
+  EXPECT_TRUE(schedule.empty());
+}
+
+TEST_F(ScheduleTest, PopBlocking) {
+  schedule.Push(1, start_time + chr::milliseconds(3));
+  EXPECT_FALSE(schedule.PopIfDue(&out));
+
+  schedule.PopBlocking(&out);
+  EXPECT_EQ(out, 1);
+  EXPECT_GE(now(), start_time + chr::milliseconds(3));
+  EXPECT_TRUE(schedule.empty());
+}
+
+TEST_F(ScheduleTest, RemoveIf) {
+  schedule.Push(1, start_time);
+  schedule.Push(2, now() + chr::minutes(1));
+  EXPECT_TRUE(schedule.RemoveIf(&out, [](const int v) { return v == 1; }));
+  EXPECT_EQ(out, 1);
+  EXPECT_TRUE(schedule.RemoveIf(&out, [](const int v) { return v == 2; }));
+  EXPECT_EQ(out, 2);
+  EXPECT_TRUE(schedule.empty());
+}
+
+TEST_F(ScheduleTest, Ordering) {
+  schedule.Push(11, start_time + chr::milliseconds(5));
+  schedule.Push(1, start_time);
+  schedule.Push(2, start_time);
+  schedule.Push(9, start_time + chr::milliseconds(2));
+  schedule.Push(3, start_time);
+  schedule.Push(10, start_time + chr::milliseconds(3));
+  schedule.Push(12, start_time + chr::milliseconds(5));
+  schedule.Push(4, start_time);
+  schedule.Push(5, start_time);
+  schedule.Push(6, start_time);
+  schedule.Push(8, start_time + chr::milliseconds(1));
+  schedule.Push(7, start_time);
+
+  std::vector<int> values;
+  const auto append = [&] {
+    values.push_back(0);
+    schedule.PopBlocking(&values.back());
+  };
+  while (!schedule.empty()) {
+    append();
+  }
+  const std::vector<int> expected = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12};
+  EXPECT_EQ(values, expected);
+}
+
+TEST_F(ScheduleTest, AddingEntryUnblocksEmptyQueue) {
+  const auto future = std::async(std::launch::async, [&] {
+    ASSERT_FALSE(schedule.PopIfDue(&out));
+    schedule.PopBlocking(&out);
+    EXPECT_EQ(out, 1);
+  });
+
+  std::this_thread::sleep_for(chr::milliseconds(5));
+  schedule.Push(1, start_time);
+  ABORT_ON_TIMEOUT(future);
+}
+
+TEST_F(ScheduleTest, PopBlockingUnblocksOnNewPastDueEntries) {
+  const auto far_away = start_time + chr::seconds(10);
+  schedule.Push(5, far_away);
+
+  const auto future = std::async(std::launch::async, [&] {
+    ASSERT_FALSE(schedule.PopIfDue(&out));
+    schedule.PopBlocking(&out);
+    EXPECT_EQ(out, 3);
+  });
+
+  std::this_thread::sleep_for(chr::milliseconds(5));
+  schedule.Push(3, start_time);
+  ABORT_ON_TIMEOUT(future);
+}
+
+TEST_F(ScheduleTest, PopBlockingAdjustsWaitTimeOnNewSoonerEntries) {
+  const auto far_away = start_time + chr::seconds(10);
+  schedule.Push(5, far_away);
+
+  const auto future = std::async(std::launch::async, [&] {
+    ASSERT_FALSE(schedule.PopIfDue(&out));
+    schedule.PopBlocking(&out);
+    EXPECT_EQ(out, 3);
+    // Make sure schedule hasn't been waiting longer than necessary.
+    EXPECT_LE(now(), far_away);
```

Comment:
Should this be EXPECT_LT? (Since it seems like now() should be roughly start_time + 100ms.)

Once you start injecting a fake clock, you could probably even do precisely this: `EXPECT_EQ(now(), start_time+100ms)`